velox/common/serialization/tests/SerializableTest.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/common/serialization/Serializable.h"
+#include <gtest/gtest.h>
+#include "folly/json.h"
+
+using namespace ::facebook::velox;
+
+namespace {
+TEST(SerializableTest, options) {
+  auto opts = getSerializationOptions();
+
+  // A very large negative number that is out of folly Json integer bound.
+  // With opts.double_fallback enabled, it should be correctly handled as double
+  // type.
+  auto largeNumber = folly::parseJson("-21820650861507478000", opts);
+  EXPECT_TRUE(largeNumber.isDouble());
+  EXPECT_EQ(largeNumber, -21820650861507478000.0);
+}
+
+class Dot : public ISerializable {
+ public:
+  Dot(std::string color, int32_t radius)
+      : color_{std::move(color)}, radius_{radius} {}
+
+  const std::string& color() const {
+    return color_;
+  }
+
+  int32_t radius() const {
+    return radius_;
+  }
+
+  folly::dynamic serialize() const override {
+    folly::dynamic obj = folly::dynamic::object();
+
+    obj["name"] = "Dot";
+    obj["color"] = color_;
+    obj["radius"] = radius_;
+    return obj;
+  }
+
+  static std::shared_ptr<Dot> create(const folly::dynamic& obj, void* context) {
+    VELOX_CHECK_NOT_NULL(context);
+    int32_t multiplier = *static_cast<int32_t*>(context);
+
+    return std::make_shared<Dot>(
+        obj["color"].asString(), obj["radius"].asInt() * multiplier);
+  }
+
+ private:
+  std::string color_;
+  int32_t radius_;
+};
+
+TEST(SerializableTest, context) {
+  DeserializationWithContextRegistryForSharedPtr().Register("Dot", Dot::create);
+
+  Dot dot{"yellow", 10};
+  auto serialized = dot.serialize();
+
+  // Deserialize with different multipliers for radius passed via 'context'.
+  int32_t multiplier = 1;
+  auto copy = ISerializable::deserialize<Dot>(serialized, &multiplier);
+  ASSERT_EQ(copy->color(), dot.color());
```

Comment:
don't objects get a default `operator=()` that lets you simply write

```
ASSERT_EQ(copy, dot);
```

?

Reply:
This doesn't work out of the box, but I added operator== and operator* to simplify the test. Thank you for the suggestion.

```
ASSERT_EQ(*copy, dot);
ASSERT_EQ(*copy, dot * 10);
```